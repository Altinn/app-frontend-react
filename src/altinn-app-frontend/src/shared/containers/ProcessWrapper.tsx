import React from 'react';

import {
  useAppSelector,
  useInstanceIdParams,
  useProcess,
} from 'src/common/hooks';
import { useApiErrorCheck } from 'src/common/hooks/useApiErrorCheck';
import UnknownError from 'src/features/instantiate/containers/UnknownError';
import { ProcessComponent } from 'src/shared/components/ProcessComponent';
import Presentation from 'src/shared/containers/Presentation';
import { InstanceDataActions } from 'src/shared/resources/instanceData/instanceDataSlice';

import {
  AltinnContentIconFormData,
  AltinnContentLoader,
} from 'altinn-shared/components';

const ProcessWrapper = () => {
  const instantiating = useAppSelector(
    (state) => state.instantiation.instantiating,
  );
  const isLoading = useAppSelector((state) => state.isLoading.dataTask);
  const { hasApiErrors } = useApiErrorCheck();
  const { dispatch, process, appOwner, appName } = useProcess();
  const { taskType } = process;

  const instanceId = useAppSelector((state) => state.instantiation.instanceId);
  const instanceIdFromUrl = useInstanceIdParams()?.instanceId;
  window['instanceId'] = instanceIdFromUrl;

  React.useEffect(() => {
    if (!instantiating && !instanceId) {
      dispatch(
        InstanceDataActions.get({
          instanceId: instanceIdFromUrl,
        }),
      );
    }
  }, [instantiating, instanceId, dispatch, instanceIdFromUrl]);
  if (hasApiErrors) {
    return <UnknownError />;
  }

  if (!process?.taskType) {
    return null;
  }
  return (
    <Presentation
      header={appName}
      appOwner={appOwner}
      type={taskType}
    >
      <ProcessComponent
        taskType={taskType}
        loading={isLoading}
      />
      {isLoading && (
        <div style={{ marginTop: '2.5rem' }}>
          <AltinnContentLoader
            width='100%'
            height={700}
          >
            <AltinnContentIconFormData />
          </AltinnContentLoader>
        </div>
      )}
    </Presentation>
  );
};

export default ProcessWrapper;
