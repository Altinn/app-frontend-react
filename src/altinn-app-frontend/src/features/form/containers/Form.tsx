import React, { useEffect } from 'react';
import { useLocation, useNavigate, useParams } from 'react-router-dom';

import Grid from '@material-ui/core/Grid';

import { useAppDispatch, useAppSelector } from 'src/common/hooks';
import ErrorReport from 'src/components/message/ErrorReport';
import { SummaryComponent } from 'src/components/summary/SummaryComponent';
import MessageBanner from 'src/features/form/components/MessageBanner';
import { DisplayGroupContainer } from 'src/features/form/containers/DisplayGroupContainer';
import { mapGroupComponents } from 'src/features/form/containers/formUtils';
import { GroupContainer } from 'src/features/form/containers/GroupContainer';
import { PanelGroupContainer } from 'src/features/form/containers/PanelGroupContainer';
import { FormLayoutActions } from 'src/features/form/layout/formLayoutSlice';
import { ReadyForPrint } from 'src/shared/components/ReadyForPrint';
import {
  extractBottomButtons,
  hasRequiredFields,
  topLevelComponents,
} from 'src/utils/formLayout';
import { renderGenericComponent as RenderGenericComponent } from 'src/utils/layout';
import {
  getFormHasErrors,
  missingFieldsInLayoutValidations,
} from 'src/utils/validation';
import type {
  ILayout,
  ILayoutComponent,
  ILayoutGroup,
} from 'src/features/form/layout';
import type { IUpdateCurrentView } from 'src/features/form/layout/formLayoutTypes';

export function renderLayoutComponent(
  layoutComponent: ILayoutComponent | ILayoutGroup,
  layout: ILayout,
) {
  switch (layoutComponent.type) {
    case 'Group': {
      return RenderLayoutGroup(layoutComponent, layout);
    }
    case 'Summary': {
      return (
        <SummaryComponent
          key={layoutComponent.id}
          {...layoutComponent}
        />
      );
    }
    default: {
      return (
        <RenderGenericComponent
          key={layoutComponent.id}
          {...layoutComponent}
        />
      );
    }
  }
}

function RenderLayoutGroup(
  layoutGroup: ILayoutGroup,
  layout: ILayout,
): JSX.Element {
  const groupComponents = mapGroupComponents(layoutGroup, layout);

  const isRepeatingGroup = layoutGroup.maxCount > 1;
  if (isRepeatingGroup) {
    return (
      <GroupContainer
        container={layoutGroup}
        id={layoutGroup.id}
        key={layoutGroup.id}
        components={groupComponents}
      />
    );
  }

  const isPanel = layoutGroup.panel;
  if (isPanel) {
    return (
      <PanelGroupContainer
        components={groupComponents}
        container={layoutGroup}
        key={layoutGroup.id}
      />
    );
  }

  //treat as regular components
  return (
    <DisplayGroupContainer
      key={layoutGroup.id}
      container={layoutGroup}
      components={groupComponents}
      renderLayoutComponent={renderLayoutComponent}
    />
  );
}

export function Form() {
  const dispatch = useAppDispatch();
  const location = useLocation();
  const viewState = location.state as IUpdateCurrentView;
  const params = useParams();
  const navigate = useNavigate();
  const [pageId] = params['*']?.split('/') || [''];
  const currentView = useAppSelector(
    (state) => state.formLayout.uiConfig.currentView,
  );
  const layout = useAppSelector(
    (state) => state.formLayout.layouts[currentView],
  );
  const language = useAppSelector((state) => state.language.language);
  const validations = useAppSelector(
    (state) => state.formValidations.validations,
  );
  const hasErrors = useAppSelector((state) =>
    getFormHasErrors(state.formValidations.validations),
  );

  const requiredFieldsMissing = React.useMemo(() => {
    if (validations && validations[currentView]) {
      return missingFieldsInLayoutValidations(
        validations[currentView],
        language,
      );
    }

    return false;
  }, [currentView, language, validations]);

  const [mainComponents, errorReportComponents] = React.useMemo(() => {
    if (!layout) {
      return [[], []];
    }
    const topLevel = topLevelComponents(layout);
    return hasErrors ? extractBottomButtons(topLevel) : [topLevel, []];
  }, [layout, hasErrors]);
  // handle root page
  useEffect(() => {
    if (!pageId) {
      if (currentView) {
        navigate(currentView, { replace: true });
      }
    }
    if (pageId !== currentView && hasErrors) {
      navigate(-1);
    }
  });
  if (viewState?.newView && viewState.newView !== currentView) {
    if (currentView !== pageId) {
      if (!hasErrors) {
        dispatch(FormLayoutActions.updateCurrentView(viewState));
      }
    }
  }
  if (!layout) {
    return <div>404</div>;
  }
  return (
    layout && (
      <>
        {hasRequiredFields(layout) && (
          <MessageBanner
            language={language}
            error={requiredFieldsMissing}
            messageKey={'form_filler.required_description'}
          />
        )}
        <Grid
          container={true}
          spacing={3}
          alignItems='flex-start'
        >
          {mainComponents.map((component) =>
            renderLayoutComponent(component, layout),
          )}
          <ErrorReport components={errorReportComponents} />
        </Grid>
        <ReadyForPrint />
      </>
    )
  );
}
