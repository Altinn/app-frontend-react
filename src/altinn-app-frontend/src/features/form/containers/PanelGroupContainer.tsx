import { Panel } from '@altinn/altinn-design-system';
import { Grid, makeStyles } from '@material-ui/core';
import React from 'react';
import { useAppSelector } from 'src/common/hooks';
import { makeGetHidden } from 'src/selectors/getLayoutData';
import { getTextFromAppOrDefault } from 'src/utils/textResource';
import { ILayoutComponent, ILayoutGroup } from '../layout';
import { renderLayoutComponent } from './Form';

const useStyles = makeStyles({
  groupContainer: {
    paddingBottom: 38,
  },
});

export interface IPanelGroupContainerProps {
  container: ILayoutGroup;
  components: (ILayoutComponent | ILayoutGroup)[];
}

export function PanelGroupContainer({ container, components }: IPanelGroupContainerProps) {
  const classes = useStyles();
  const GetHiddenSelector = makeGetHidden();
  const layout = useAppSelector(state => state.formLayout.layouts[state.formLayout.uiConfig.currentView]);
  const hidden = useAppSelector(state => GetHiddenSelector(state, { id: container.id }));
  const title = useAppSelector(state => getTextFromAppOrDefault(container.textResourceBindings?.title, state.textResources.resources, state.language.language, [], true));

  if (hidden) {
    return null;
  }

  return (
    <Grid
      container={true}
      item={true}
      className={classes.groupContainer}
      spacing={3}
      alignItems='flex-start'
      data-testid='panel-group-container'
    >
        <Panel title={title}>
          {components.map((component) => {
            return renderLayoutComponent(component, layout);
          })}
        </Panel >
    </Grid>
  );
}
