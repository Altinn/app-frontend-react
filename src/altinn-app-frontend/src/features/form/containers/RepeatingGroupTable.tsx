import React, { useState } from 'react';

import { createTheme, Grid, makeStyles, TableCell, TableRow, useMediaQuery } from '@material-ui/core';

import { ExprDefaultsForGroup } from 'src/features/expressions';
import { useExpressions } from 'src/features/expressions/useExpressions';
import AltinnMobileTableItem from 'src/features/form/containers/AltinnMobileTableItem';
import { RepeatingGroupsEditContainer } from 'src/features/form/containers/RepeatingGroupsEditContainer';
import { RepeatingGroupTableRow } from 'src/features/form/containers/RepeatingGroupTableRow';
import { getFormDataForComponentInRepeatingGroup, getTextResource } from 'src/utils/formComponentUtils';
import { createRepeatingGroupComponents } from 'src/utils/formLayout';
import { setupGroupComponents } from 'src/utils/layout';
import { componentHasValidations, repeatingGroupHasValidations } from 'src/utils/validation';
import type { IMobileTableItem } from 'src/features/form/containers/AltinnMobileTableItem';
import type { IFormData } from 'src/features/form/data';
import type { ILayout, ILayoutCompInput, ILayoutComponent, ILayoutGroup } from 'src/features/form/layout';
import type { IAttachments } from 'src/shared/resources/attachments';
import type { IOptions, IRepeatingGroups, ITextResource, ITextResourceBindings, IValidations } from 'src/types';

import { AltinnMobileTable, AltinnTable, AltinnTableBody, AltinnTableHeader } from 'altinn-shared/components';
import altinnAppTheme from 'altinn-shared/theme/altinnAppTheme';
import { getLanguageFromKey, getTextResourceByKey } from 'altinn-shared/utils';
import type { ILanguage } from 'altinn-shared/types';

export interface IRepeatingGroupTableProps {
  id: string;
  container: ILayoutGroup;
  components: (ILayoutComponent | ILayoutGroup)[];
  repeatingGroupIndex: number;
  repeatingGroups: IRepeatingGroups | null;
  repeatingGroupDeepCopyComponents: (ILayoutComponent | ILayoutGroup)[][];
  hiddenFields: string[];
  formData: IFormData;
  attachments: IAttachments;
  options: IOptions;
  textResources: ITextResource[];
  language: ILanguage;
  currentView: string;
  layout: ILayout | null;
  validations: IValidations;
  editIndex: number;
  setEditIndex: (index: number, forceValidation?: boolean) => void;
  onClickRemove: (groupIndex: number) => void;
  setMultiPageIndex?: (index: number) => void;
  multiPageIndex?: number;
  deleting: boolean;
  filteredIndexes?: number[] | null;
}

const theme = createTheme(altinnAppTheme);

const useStyles = makeStyles({
  editContainerInTable: {
    borderTop: `1px solid ${theme.altinnPalette.primary.blueLight}`,
    marginBottom: 0,
  },
  editContainerRow: {
    '&:hover': {
      background: 'unset !important',
    },
    '& td': {
      whiteSpace: 'normal',
    },
  },
  visuallyHidden: {
    border: 0,
    padding: 0,
    margin: 0,
    position: 'absolute',
    height: '1px',
    width: '1px',
    overflow: 'hidden',
    clip: 'rect(1px 1px 1px 1px)',
    clipPath: 'inset(50%)',
    whiteSpace: 'nowrap',
  },
});

function getEditButtonText(
  language: ILanguage,
  isEditing: boolean,
  textResources: ITextResource[],
  textResourceBindings?: ITextResourceBindings,
): string {
  if (isEditing && textResourceBindings?.edit_button_close) {
    return getTextResourceByKey(textResourceBindings?.edit_button_close, textResources);
  } else if (!isEditing && textResourceBindings?.edit_button_open) {
    return getTextResourceByKey(textResourceBindings?.edit_button_open, textResources);
  }

  return isEditing
    ? getLanguageFromKey('general.save_and_close', language)
    : getLanguageFromKey('general.edit_alt', language);
}

function getTableTitle(textResourceBindings: ITextResourceBindings) {
  if (textResourceBindings.tableTitle) {
    return textResourceBindings.tableTitle;
  }
  if (textResourceBindings.title) {
    return textResourceBindings.title;
  }
  return '';
}

function getTextAlignment(component: ILayoutComponent): 'left' | 'center' | 'right' {
  const formatting = (component as ILayoutCompInput).formatting;
  if (formatting && formatting.align) {
    return formatting.align;
  }
  if (formatting && formatting.number) {
    return 'right';
  }
  return 'left';
}

export function RepeatingGroupTable({
  id,
  container,
  components,
  repeatingGroupIndex,
  repeatingGroupDeepCopyComponents,
  editIndex,
  formData,
  attachments,
  options,
  textResources,
  currentView,
  hiddenFields,
  language,
  layout,
  repeatingGroups,
  validations,
  setEditIndex,
  onClickRemove,
  setMultiPageIndex,
  multiPageIndex,
  deleting,
  filteredIndexes,
}: IRepeatingGroupTableProps): JSX.Element {
  const classes = useStyles();
  const mobileView = useMediaQuery('(max-width:992px)'); // breakpoint on altinn-modal

  const edit = useExpressions(container.edit, {
    forComponentId: id,
    defaults: ExprDefaultsForGroup.edit,
  });

  const tableHeaderComponentIds = container.tableHeaders || components.map((c) => c.baseComponentId || c.id) || [];

  const componentsDeepCopy: ILayoutComponent[] = JSON.parse(JSON.stringify(components));
  const tableComponents = componentsDeepCopy.filter((component: ILayoutComponent) => {
    const childId = component.baseComponentId || component.id;
    return tableHeaderComponentIds.includes(childId);
  });

  const componentTextResourceBindings: ITextResourceBindings[] = [];
  tableComponents.forEach((component) => {
    componentTextResourceBindings.push(component.textResourceBindings as ITextResourceBindings);
  });

  const componentTextResourceBindingsResolved = useExpressions(componentTextResourceBindings);

  const showTableHeader = repeatingGroupIndex > -1 && !(repeatingGroupIndex == 0 && editIndex == 0);
  const [displayDeleteColumn, setDisplayDeleteColumn] = useState(edit?.deleteButton);
  const [popoverPanelIndex, setPopoverPanelIndex] = useState(-1);
  const [popoverOpen, setPopoverOpen] = useState(false);

  const getFormDataForComponent = (component: ILayoutComponent | ILayoutGroup, index: number): string => {
    return getFormDataForComponentInRepeatingGroup(
      formData,
      attachments,
      component,
      index,
      container.dataModelBindings?.group,
      textResources,
      options,
      repeatingGroups,
    );
  };

  const onOpenChange = (index: number) => {
    if (index == popoverPanelIndex && popoverOpen) {
      setPopoverPanelIndex(-1);
    } else {
      setPopoverPanelIndex(index);
    }
  };

  const handlePopoverDeleteClick = (index: number) => {
    return () => {
      onClickRemove(index);
      onOpenChange(index);
      setPopoverOpen(false);
    };
  };

  const handleDeleteClick = (index: number) => {
    if (edit?.alertOnDelete) {
      onOpenChange(index);
    } else {
      onClickRemove(index);
    }
  };

  const handleEditClick = (groupIndex: number) => {
    if (groupIndex === editIndex) {
      setEditIndex(-1);
    } else {
      setEditIndex(groupIndex);
    }
  };

  const childElementHasErrors = (element: ILayoutGroup | ILayoutComponent, index: number) => {
    if (element.type === 'Group') {
      return childGroupHasErrors(element as ILayoutGroup, index);
    }
    return componentHasValidations(validations, currentView, `${element.id}`);
  };

  const childGroupHasErrors = (childGroup: ILayoutGroup, index: number) => {
    if (!repeatingGroups || !layout) {
      return;
    }

    const childGroupIndex = repeatingGroups[childGroup.id]?.index;
    const childGroupComponents = layout.filter((childElement) => childGroup.children?.indexOf(childElement.id) > -1);
    const childRenderComponents = setupGroupComponents(
      childGroupComponents,
      childGroup.dataModelBindings?.group,
      index,
    );
    const deepCopyComponents = createRepeatingGroupComponents(
      childGroup,
      childRenderComponents,
      childGroupIndex,
      textResources,
      hiddenFields,
    );
    return repeatingGroupHasValidations(
      childGroup,
      deepCopyComponents,
      validations,
      currentView,
      repeatingGroups,
      layout,
      hiddenFields,
    );
  };

  const renderRepeatingGroupsEditContainer = () => {
    return (
      editIndex >= 0 && (
        <RepeatingGroupsEditContainer
          className={classes.editContainerInTable}
          container={container}
          editIndex={editIndex}
          setEditIndex={setEditIndex}
          repeatingGroupIndex={repeatingGroupIndex}
          id={id}
          language={language}
          textResources={textResources}
          layout={layout}
          repeatingGroupDeepCopyComponents={repeatingGroupDeepCopyComponents}
          hideSaveButton={edit?.saveButton === false}
          multiPageIndex={multiPageIndex}
          setMultiPageIndex={setMultiPageIndex}
          showSaveAndNextButton={edit?.saveAndNextButton === true}
          filteredIndexes={filteredIndexes}
        />
      )
    );
  };

  return (
    <Grid
      container={true}
      item={true}
      data-testid={`group-${id}`}
      id={`group-${id}`}
    >
      {!mobileView && (
        <AltinnTable id={`group-${id}-table`}>
          {showTableHeader && (
            <AltinnTableHeader
              showBorder={editIndex !== 0}
              id={`group-${id}-table-header`}
            >
              <TableRow>
                {tableComponents.map((component: ILayoutComponent, tableComponentIndex: number) => (
                  <TableCell
                    align={getTextAlignment(component)}
                    key={component.id}
                  >
                    {getTextResource(
                      getTableTitle(componentTextResourceBindingsResolved[tableComponentIndex]),
                      textResources,
                    )}
                  </TableCell>
                ))}
                <TableCell style={{ width: '185px', padding: 0, paddingRight: '10px' }}>
                  <span className={classes.visuallyHidden}>{getLanguageFromKey('general.edit', language)}</span>
                </TableCell>
                {displayDeleteColumn && (
                  <TableCell style={{ width: '120px', padding: 0 }}>
                    <span className={classes.visuallyHidden}>{getLanguageFromKey('general.delete', language)}</span>
                  </TableCell>
                )}
              </TableRow>
            </AltinnTableHeader>
          )}
          <AltinnTableBody id={`group-${id}-table-body`}>
            {repeatingGroupIndex >= 0 &&
              [...Array(repeatingGroupIndex + 1)].map((_x: any, index: number) => {
                const rowHasErrors = repeatingGroupDeepCopyComponents[index].some(
                  (component: ILayoutComponent | ILayoutGroup) => {
                    return childElementHasErrors(component, index);
                  },
                );

                // Check if filter is applied and includes specified index.
                if (filteredIndexes && !filteredIndexes.includes(index)) {
                  return null;
                }

                return (
                  <React.Fragment key={index}>
                    <RepeatingGroupTableRow
                      id={id}
                      container={container}
                      components={components}
                      repeatingGroups={repeatingGroups}
                      formData={formData}
                      attachments={attachments}
                      options={options}
                      textResources={textResources}
                      language={language}
                      editIndex={editIndex}
                      setEditIndex={setEditIndex}
                      onClickRemove={onClickRemove}
                      deleting={deleting}
                      index={index}
                      rowHasErrors={rowHasErrors}
                      tableComponents={tableComponents}
                      setDisplayDeleteColumn={setDisplayDeleteColumn}
                    />
                    {editIndex === index && (
                      <TableRow
                        key={`edit-container-${index}`}
                        className={classes.editContainerRow}
                      >
                        <TableCell
                          style={{ padding: 0, borderBottom: 0 }}
                          colSpan={displayDeleteColumn ? tableComponents.length + 2 : tableComponents.length + 1}
                        >
                          {renderRepeatingGroupsEditContainer()}
                        </TableCell>
                      </TableRow>
                    )}
                  </React.Fragment>
                );
              })}
          </AltinnTableBody>
        </AltinnTable>
      )}
      {mobileView && (
        <AltinnMobileTable
          id={`group-${id}-table`}
          showBorder={showTableHeader && editIndex !== 0}
        >
          {repeatingGroupIndex >= 0 &&
            [...Array(repeatingGroupIndex + 1)].map((_x: any, index: number) => {
              const rowHasErrors = repeatingGroupDeepCopyComponents[index].some(
                (component: ILayoutComponent | ILayoutGroup) => {
                  return childElementHasErrors(component, index);
                },
              );
              const items: IMobileTableItem[] = tableComponents.map(
                (component: ILayoutComponent, tableComponentIndex: number) => ({
                  key: component.id,
                  label: getTextResource(
                    getTableTitle(componentTextResourceBindingsResolved[tableComponentIndex]),
                    textResources,
                  ),
                  value: getFormDataForComponent(component, index),
                }),
              );
              return (
                <React.Fragment key={index}>
                  <AltinnMobileTableItem
                    key={`mobile-table-item-${index}`}
                    tableItemIndex={index}
                    container={container}
                    textResources={textResources}
                    language={language}
                    items={items}
                    valid={!rowHasErrors}
                    editIndex={editIndex}
                    onEditClick={() => handleEditClick(index)}
                    getEditButtonText={getEditButtonText}
                    deleteFunctionality={{
                      onDeleteClick: () => handleDeleteClick(index),
                      deleteButtonText: getLanguageFromKey('general.delete', language),
                      popoverPanelIndex,
                      popoverOpen,
                      setPopoverOpen,
                      onPopoverDeleteClick: handlePopoverDeleteClick,
                      onOpenChange,
                    }}
                  />
                  {editIndex === index && renderRepeatingGroupsEditContainer()}
                </React.Fragment>
              );
            })}
        </AltinnMobileTable>
      )}
    </Grid>
  );
}
