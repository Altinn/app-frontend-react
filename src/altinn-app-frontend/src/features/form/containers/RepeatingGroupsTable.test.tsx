import * as React from 'react';

import { getFormLayoutGroupMock } from '__mocks__/mocks';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { mockMediaQuery, renderWithProviders } from 'testUtils';

import { RepeatingGroupTable } from 'src/features/form/containers/RepeatingGroupTable';
import { createRepeatingGroupComponents } from 'src/utils/formLayout';
import type { IRepeatingGroupTableProps } from 'src/features/form/containers/RepeatingGroupTable';
import type {
  ILayoutComponent,
  ILayoutGroup,
  ISelectionComponentProps,
} from 'src/features/form/layout';
import type { ILayoutState } from 'src/features/form/layout/formLayoutSlice';
import type { IAttachments } from 'src/shared/resources/attachments';
import type { IOption, ITextResource } from 'src/types';

import type { ILanguage } from 'altinn-shared/types';

const user = userEvent.setup();

let mockContainer: ILayoutGroup;
let mockLanguage: ILanguage;
let mockTextResources: ITextResource[];
let mockAttachments: IAttachments;
let mockComponents: ILayoutComponent[];
let mockTestLayout: ILayoutState;
let mockTestCurrentView: string;
let mockData: any;
let mockOptions: IOption[];
let repeatingGroupIndex: number;
let mockRepeatingGroupDeepCopyComponents: Array<
  Array<ILayoutComponent | ILayoutGroup>
>;

describe('RepeatingGroupTable', () => {
  beforeEach(() => {
    mockOptions = [{ value: 'option.value', label: 'option.label' }];
    mockComponents = [
      {
        id: 'field1',
        type: 'Input',
        dataModelBindings: {
          simpleBinding: 'Group.prop1',
        },
        textResourceBindings: {
          title: 'Title1',
        },
        readOnly: false,
        required: false,
        disabled: false,
      },
      {
        id: 'field2',
        type: 'Input',
        dataModelBindings: {
          simpleBinding: 'Group.prop2',
        },
        textResourceBindings: {
          title: 'Title2',
        },
        readOnly: false,
        required: false,
        disabled: false,
      },
      {
        id: 'field3',
        type: 'Input',
        dataModelBindings: {
          simpleBinding: 'Group.prop3',
        },
        textResourceBindings: {
          title: 'Title3',
        },
        readOnly: false,
        required: false,
        disabled: false,
      },
      {
        id: 'field4',
        type: 'Checkboxes',
        dataModelBindings: {
          simpleBinding: 'some-group.checkboxBinding',
        },
        textResourceBindings: {
          title: 'Title4',
        },
        readOnly: false,
        required: false,
        disabled: false,
        options: mockOptions,
      } as ISelectionComponentProps,
    ];

    mockContainer = getFormLayoutGroupMock({});

    mockLanguage = {
      general: {
        delete: 'Delete',
        edit_alt: 'Edit',
      },
    };

    mockTestLayout = {
      layouts: {
        FormLayout: [].concat(mockContainer).concat(mockComponents),
      },
      uiConfig: {
        hiddenFields: [],
        repeatingGroups: {
          'mock-container-id': {
            index: 3,
          },
        },
        autoSave: false,
        currentView: 'FormLayout',
        focus: undefined,
        layoutOrder: ['FormLayout'],
      },
      error: null,
      layoutsets: null,
    };

    mockData = {
      formData: {
        'some-group[1].checkboxBinding': 'option.value',
      },
    };

    mockTextResources = [
      { id: 'option.label', value: 'Value to be shown' },
      { id: 'edit_button_open', value: 'Edit' },
      { id: 'edit_button_close', value: 'Edit' },
    ];

    mockTestCurrentView = 'FormLayout';
    repeatingGroupIndex = 3;
    mockRepeatingGroupDeepCopyComponents = createRepeatingGroupComponents(
      mockContainer,
      mockComponents,
      repeatingGroupIndex,
      mockTextResources,
    );
  });

  describe('desktop view', () => {
    const { setScreenWidth } = mockMediaQuery(992);
    beforeEach(() => {
      setScreenWidth(1337);
    });

    it('should match snapshot', () => {
      const { asFragment } = render();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should trigger onClickRemove on delete-button click', async () => {
      const onClickRemove = jest.fn();
      render({ onClickRemove: onClickRemove });

      await user.click(screen.getAllByRole('button', { name: /delete/i })[0]);

      expect(onClickRemove).toBeCalledTimes(1);
    });

    it('should trigger setEditIndex on edit-button click', async () => {
      const setEditIndex = jest.fn();
      render({ setEditIndex: setEditIndex });

      await user.click(screen.getAllByRole('button', { name: /edit/i })[0]);

      expect(setEditIndex).toBeCalledTimes(1);
    });
  });

  describe('tablet view', () => {
    const { setScreenWidth } = mockMediaQuery(992);
    beforeEach(() => {
      setScreenWidth(992);
    });

    it('should render as mobile-version for small screens', () => {
      render();

      const altinnMobileTable = screen.queryByTestId(/altinn-mobile-table/i);

      expect(altinnMobileTable).toBeInTheDocument();
    });
  });

  describe('mobile view', () => {
    const { setScreenWidth } = mockMediaQuery(768);
    beforeEach(() => {
      setScreenWidth(768);
    });

    it('should render edit and delete buttons as icons for screens smaller thnn 786px', () => {
      render();

      const iconButtonsDelete = screen.getAllByTestId(/delete-button/i);
      const iconButtonsEdit = screen.getAllByTestId(/edit-button/i);

      expect(iconButtonsDelete).toHaveLength(4);
      expect(iconButtonsDelete[0]).toContainHTML(
        `<span class="MuiIconButton-label"><i class="ai ai-trash" /></span>`,
      );
      expect(iconButtonsEdit).toHaveLength(4);
      expect(iconButtonsEdit[0]).toContainHTML(
        `<span class="MuiIconButton-label"><i class="fa fa-edit makeStyles-editIcon-60" /></span>`,
      );
    });
  });
});

const render = (props: Partial<IRepeatingGroupTableProps> = {}) => {
  const allProps = {
    container: mockContainer,
    attachments: mockAttachments,
    language: mockLanguage,
    textResources: mockTextResources,
    components: mockComponents,
    currentView: mockTestCurrentView,
    editIndex: -1,
    formData: mockData,
    hiddenFields: [],
    id: mockContainer.id,
    layout: mockTestLayout.layouts[mockTestCurrentView],
    options: {},
    repeatingGroupDeepCopyComponents: mockRepeatingGroupDeepCopyComponents,
    repeatingGroupIndex: repeatingGroupIndex,
    repeatingGroups: mockTestLayout.uiConfig.repeatingGroups,
    deleting: false,
    onClickRemove: jest.fn(),
    setEditIndex: jest.fn(),
    validations: {},
    ...props,
  } as IRepeatingGroupTableProps;

  return renderWithProviders(<RepeatingGroupTable {...allProps} />);
};
