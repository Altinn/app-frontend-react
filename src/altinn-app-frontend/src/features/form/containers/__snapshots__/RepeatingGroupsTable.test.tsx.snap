// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features > form > containers > RepeatingGroupTable.tsx should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-item"
    data-testid="group-mock-container-id"
    id="group-mock-container-id"
  >
    <div
      class="MuiGrid-root MuiTableContainer-root"
      id="group-mock-container-id-table-container"
    >
      <table
        class="MuiTable-root makeStyles-table-1"
        id="group-mock-container-id-table"
      >
        <thead
          class="MuiTableHead-root makeStyles-tableHeader-3"
          id="group-mock-container-id-table-header"
        >
          <tr
            class="MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-alignLeft"
            >
              Title1
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-alignLeft"
            >
              Title2
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-alignLeft"
            >
              Title3
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-alignLeft"
            >
              Title4
            </td>
            <td
              class="MuiTableCell-root"
            />
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root makeStyles-tableBody-4"
          id="group-mock-container-id-table-body"
        >
          <tr
            class="MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root"
            />
            <td
              class="MuiTableCell-root"
            />
            <td
              class="MuiTableCell-root"
            />
            <td
              class="MuiTableCell-root"
            />
            <td
              class="MuiTableCell-root MuiTableCell-alignRight"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root makeStyles-tableEditButton-3"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <i
                    class="fa fa-edit makeStyles-editIcon-2"
                  />
                  general.edit_alt
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root"
            />
            <td
              class="MuiTableCell-root"
            />
            <td
              class="MuiTableCell-root"
            />
            <td
              class="MuiTableCell-root"
            />
            <td
              class="MuiTableCell-root MuiTableCell-alignRight"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root makeStyles-tableEditButton-3"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <i
                    class="fa fa-edit makeStyles-editIcon-2"
                  />
                  general.edit_alt
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root"
            />
            <td
              class="MuiTableCell-root"
            />
            <td
              class="MuiTableCell-root"
            />
            <td
              class="MuiTableCell-root"
            />
            <td
              class="MuiTableCell-root MuiTableCell-alignRight"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root makeStyles-tableEditButton-3"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <i
                    class="fa fa-edit makeStyles-editIcon-2"
                  />
                  general.edit_alt
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root"
            />
            <td
              class="MuiTableCell-root"
            />
            <td
              class="MuiTableCell-root"
            />
            <td
              class="MuiTableCell-root"
            />
            <td
              class="MuiTableCell-root MuiTableCell-alignRight"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root makeStyles-tableEditButton-3"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <i
                    class="fa fa-edit makeStyles-editIcon-2"
                  />
                  general.edit_alt
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
